File: BotsSDK/korebot/src/main/java/kore/botssdk/dialogs/OptionsActionSheetFragment.java
package kore.botssdk.dialogs;

import android.app.Dialog;
import android.content.Context;
import android.content.DialogInterface;
import android.content.SharedPreferences;
import android.graphics.Color;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.FrameLayout;
import android.widget.LinearLayout;
import android.widget.ListView;
import android.widget.TextView;

import androidx.annotation.Nullable;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.google.android.material.bottomsheet.BottomSheetBehavior;
import com.google.android.material.bottomsheet.BottomSheetDialog;
import com.google.android.material.bottomsheet.BottomSheetDialogFragment;

import kore.botssdk.R;
import kore.botssdk.adapter.BottomOptionsCycleAdapter;
import kore.botssdk.application.AppControl;
import kore.botssdk.listener.ComposeFooterInterface;
import kore.botssdk.models.BotOptionsModel;
import kore.botssdk.models.BotResponse;
import kore.botssdk.viewUtils.DimensionUtil;

public class OptionsActionSheetFragment extends BottomSheetDialogFragment
{
    private BotOptionsModel model;
    private ComposeFooterInterface composeFooterInterface;
    private int dp1;
    private BottomSheetDialog bottomSheetDialog;

    @Override
    public View onCreateView(LayoutInflater inflater,
                             @Nullable ViewGroup container,
                             @Nullable Bundle savedInstanceState) {

        View view = inflater.inflate(R.layout.list_bottom_sheet, container, false);
        ListView lvMoreData = view.findViewById(R.id.lvMoreData);
        TextView tvOptionsTitle = view.findViewById(R.id.tvOptionsTitle);
        LinearLayout llCloseBottomSheet = view.findViewById(R.id.llCloseBottomSheet);
        LinearLayout llBottomLayout = view.findViewById(R.id.llBottomLayout);
        LinearLayout llTabHeader = view.findViewById(R.id.llTabHeader);
        llTabHeader.setVisibility(View.GONE);
        tvOptionsTitle.setVisibility(View.VISIBLE);
        RecyclerView rvViewMore = view.findViewById(R.id.rvMoreData);
        SharedPreferences sharedPreferences = requireActivity().getSharedPreferences(BotResponse.THEME_NAME, Context.MODE_PRIVATE);

        if(sharedPreferences != null)
            llBottomLayout.setBackgroundColor(Color.parseColor(sharedPreferences.getString(BotResponse.WIDGET_BG_COLOR, "#FFFFFF")));

        rvViewMore.setLayoutManager(new LinearLayoutManager(requireActivity(), LinearLayoutManager.VERTICAL, false));
        rvViewMore.setVisibility(View.VISIBLE);
        lvMoreData.setVisibility(View.GONE);
        this.dp1 = (int) DimensionUtil.dp1;
        BottomOptionsCycleAdapter bottomOptionsCycleAdapter;

        if (rvViewMore.getAdapter() == null) {
            bottomOptionsCycleAdapter = new BottomOptionsCycleAdapter(model.getTasks());
            rvViewMore.setAdapter(bottomOptionsCycleAdapter);
            bottomOptionsCycleAdapter.setComposeFooterInterface(composeFooterInterface);
            bottomOptionsCycleAdapter.setContext(requireActivity());
        } else {
            bottomOptionsCycleAdapter = (BottomOptionsCycleAdapter) rvViewMore.getAdapter();
        }

        bottomOptionsCycleAdapter.setBotListModelArrayList(bottomSheetDialog, model.getTasks());
        bottomOptionsCycleAdapter.notifyDataSetChanged();

        if(model.getHeading() != null)
        {
            tvOptionsTitle.setText(model.getHeading());

            if(sharedPreferences != null)
                tvOptionsTitle.setTextColor(Color.parseColor(sharedPreferences.getString(BotResponse.WIDGET_TXT_COLOR, "#000000")));
        }

        llCloseBottomSheet.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                if(bottomSheetDialog != null)
                    bottomSheetDialog.dismiss();
            }
        });

        return view;
    }

    public void setComposeFooterInterface(ComposeFooterInterface composeFooterInterface) {
        this.composeFooterInterface = composeFooterInterface;
    }


    @Override
    public Dialog onCreateDialog(Bundle savedInstanceState) {
        bottomSheetDialog = (BottomSheetDialog) super.onCreateDialog(savedInstanceState);
        bottomSheetDialog.setOnShowListener(new DialogInterface.OnShowListener() {
            @Override
            public void onShow(DialogInterface dialogInterface) {
                BottomSheetDialog d = (BottomSheetDialog) dialogInterface;
                FrameLayout bottomSheet = d.findViewById(com.google.android.material.R.id.design_bottom_sheet);

                bottomSheet.getLayoutParams().height = (int) (AppControl.getInstance(getContext()).getDimensionUtil().screenHeight)-(50 * dp1);
                BottomSheetBehavior bottomSheetBehavior = BottomSheetBehavior.from(bottomSheet);
                bottomSheetBehavior.setPeekHeight((int) ((AppControl.getInstance(getContext()).getDimensionUtil().screenHeight) -(50 * dp1)));
            }
        });

        // Do something with your dialog like setContentView() or whatever
        return bottomSheetDialog;
    }

    public void setData(BotOptionsModel taskTemplateModel) {
        model = taskTemplateModel;
    }
}

File: BotSdk/ui/src/main/java/com/kore/ui/dialog/OptionsActionSheetFragment.kt
package com.kore.ui.dialog

import android.annotation.SuppressLint
import android.app.Dialog
import android.content.Context
import android.content.DialogInterface
import android.graphics.Color
import android.os.Bundle
import android.util.DisplayMetrics
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.FrameLayout
import android.widget.LinearLayout
import android.widget.TextView
import androidx.core.view.isVisible
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.google.android.flexbox.FlexDirection
import com.google.android.flexbox.FlexboxLayoutManager
import com.google.android.flexbox.JustifyContent
import com.google.android.material.bottomsheet.BottomSheetBehavior
import com.google.android.material.bottomsheet.BottomSheetDialog
import com.google.android.material.bottomsheet.BottomSheetDialogFragment
import com.kore.common.event.UserActionEvent
import com.kore.extensions.dpToPx
import com.kore.model.constants.BotResponseConstants
import com.kore.network.api.responsemodels.branding.BotBrandingModel
import com.kore.network.api.responsemodels.branding.BrandingQuickStartButtonActionModel
import com.kore.ui.R
import com.kore.ui.adapters.BottomMenuOptionsAdapter
import com.kore.ui.adapters.WelcomeStarterButtonsAdapter

@SuppressLint("UnknownNullness")
class OptionsActionSheetFragment : BottomSheetDialogFragment() {
    private var models: ArrayList<BrandingQuickStartButtonActionModel>? = null
    private var isFromListMenu: Boolean = false
    private var bottomSheetDialog: BottomSheetDialog? = null
    private var botBrandingModel: BotBrandingModel? = null
    private var actionEvent: (event: UserActionEvent) -> Unit = {}

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        val view = inflater.inflate(R.layout.menus_bottom_sheet, container, false)
        val tvOptionsTitle = view.findViewById<TextView>(R.id.tvOptionsTitle)
        val tvMenuTitle = view.findViewById<TextView>(R.id.tvMenuTitle)
        val llCloseBottomSheet = view.findViewById<LinearLayout>(R.id.llCloseBottomSheet)
        val llBottomLayout = view.findViewById<LinearLayout>(R.id.llBottomLayout)

        val llTabHeader = view.findViewById<LinearLayout>(R.id.llTabHeader)
        val rvViewMore = view.findViewById<RecyclerView>(R.id.rvMoreData)
        val rvQuickData = view.findViewById<RecyclerView>(R.id.rvQuickData)

        llTabHeader.isVisible = false
        tvOptionsTitle.isVisible = true
        rvQuickData.isVisible = false
        tvMenuTitle.isVisible = false

        val sharedPreferences = requireActivity().getSharedPreferences(BotResponseConstants.THEME_NAME, Context.MODE_PRIVATE)

        if (sharedPreferences != null)
            llBottomLayout.setBackgroundColor(Color.parseColor(sharedPreferences.getString(BotResponseConstants.WIDGET_BG_COLOR, "#FFFFFF")))

        val layoutManager = FlexboxLayoutManager(requireActivity())
        layoutManager.flexDirection = FlexDirection.ROW
        layoutManager.justifyContent = JustifyContent.FLEX_START
        rvViewMore.layoutManager = layoutManager
        rvViewMore.layoutManager = LinearLayoutManager(requireActivity(), LinearLayoutManager.VERTICAL, false)

        val bottomOptionsCycleAdapter = BottomMenuOptionsAdapter(models!!)
        bottomOptionsCycleAdapter.setBotListModelArrayList(bottomSheetDialog)
        bottomOptionsCycleAdapter.setActionEvent(actionEvent)
        rvViewMore.adapter = bottomOptionsCycleAdapter

//        bottomOptionsCycleAdapter.notifyItemRangeChanged(0, (models!!.size - 1))

        if (botBrandingModel != null) {
            if (!botBrandingModel?.welcomeScreen?.starterBox?.quickStartButtons?.buttons.isNullOrEmpty()) {
                if (botBrandingModel!!.welcomeScreen.starterBox.quickStartButtons.style.equals(BotResponseConstants.TEMPLATE_TYPE_LIST, ignoreCase = true)) {
                    layoutManager.flexDirection = FlexDirection.COLUMN
                    layoutManager.justifyContent = JustifyContent.FLEX_START
                    rvQuickData.layoutManager = layoutManager

                    val quickRepliesAdapter = WelcomeStarterButtonsAdapter(requireActivity(), BotResponseConstants.TEMPLATE_TYPE_LIST)
                    quickRepliesAdapter.setWelcomeStarterButtonsArrayList(botBrandingModel!!.welcomeScreen.starterBox.quickStartButtons.buttons)
                    rvQuickData.adapter = quickRepliesAdapter
                } else {
                    layoutManager.flexDirection = FlexDirection.ROW
                    layoutManager.justifyContent = JustifyContent.FLEX_START
                    rvQuickData.layoutManager = layoutManager

                    val quickRepliesAdapter = WelcomeStarterButtonsAdapter(requireActivity(), BotResponseConstants.TEMPLATE_TYPE_CAROUSEL)
                    quickRepliesAdapter.setWelcomeStarterButtonsArrayList(botBrandingModel!!.welcomeScreen.starterBox.quickStartButtons.buttons)
                    rvQuickData.adapter = quickRepliesAdapter
                }
            }
        }

        tvOptionsTitle.text = requireActivity().resources.getString(R.string.menu)

        if (sharedPreferences != null) tvOptionsTitle.setTextColor(
            Color.parseColor(sharedPreferences.getString(BotResponseConstants.WIDGET_TXT_COLOR, "#000000"))
        )

        llCloseBottomSheet.setOnClickListener { v: View? ->
            if (bottomSheetDialog != null) bottomSheetDialog!!.dismiss()
        }

        return view
    }

    override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
        bottomSheetDialog = super.onCreateDialog(savedInstanceState) as BottomSheetDialog
        if (bottomSheetDialog?.window != null) {
            bottomSheetDialog?.window?.attributes?.windowAnimations = R.style.DialogAnimation
        }

        bottomSheetDialog?.setOnShowListener { dialogInterface: DialogInterface ->
            val d = dialogInterface as BottomSheetDialog
            val bottomSheet = d.findViewById<FrameLayout>(com.google.android.material.R.id.design_bottom_sheet)
            if (bottomSheet != null) {
                val displayMetrics = DisplayMetrics()
                requireActivity().windowManager.defaultDisplay.getMetrics(displayMetrics)
                bottomSheet.layoutParams.height = displayMetrics.heightPixels
                val bottomSheetBehavior = BottomSheetBehavior.from(bottomSheet)
                bottomSheetBehavior.peekHeight = (models!!.size.dpToPx(bottomSheetDialog?.context!!) * 125).toInt()
                bottomSheetBehavior.isDraggable = false
            }
        }

        // Do something with your dialog like setContentView() or whatever
        return bottomSheetDialog!!
    }

    fun setData(taskTemplateModel: ArrayList<BrandingQuickStartButtonActionModel>?) {
        models = taskTemplateModel
    }

    fun setData(botOptionsModel: ArrayList<BrandingQuickStartButtonActionModel>?, isFromListMenu: Boolean) {
        models = botOptionsModel
        this.isFromListMenu = isFromListMenu
    }

    fun setBrandingModel(botBrandingModel: BotBrandingModel?) {
        this.botBrandingModel = botBrandingModel
    }

    fun setActionEvent(actionEvent: (event: UserActionEvent) -> Unit) {
        this.actionEvent = actionEvent
    }
}
